File: src/main/java/buttondevteam/discordplugin/DiscordSender.java
package buttondevteam.discordplugin;

import org.bukkit.Bukkit;
import org.bukkit.Server;
import org.bukkit.command.CommandSender;
import org.bukkit.permissions.PermissibleBase;
import org.bukkit.permissions.Permission;
import org.bukkit.permissions.PermissionAttachment;
import org.bukkit.permissions.PermissionAttachmentInfo;
import org.bukkit.plugin.Plugin;
import sx.blah.discord.handle.obj.IChannel;
import sx.blah.discord.handle.obj.IUser;

import java.util.Set;

public class DiscordSender extends DiscordSenderBase implements CommandSender {
	private PermissibleBase perm = new PermissibleBase(this);

	private String name = null;

	public DiscordSender(IUser user, IChannel channel) {
		super(user, channel);
	}

	public DiscordSender(IUser user, IChannel channel, String name) {
		super(user, channel);
		this.name = name;
	}

	@Override
	public boolean isPermissionSet(String name) {
		return perm.isPermissionSet(name);
	}

	@Override
	public boolean isPermissionSet(Permission perm) {
		return this.perm.isPermissionSet(perm);
	}

	@Override
	public boolean hasPermission(String name) {
		if (name.contains("essentials") && !name.equals("essentials.list"))
			return false;
		return perm.hasPermission(name);
	}

	@Override
	public boolean hasPermission(Permission perm) {
		return this.perm.hasPermission(perm);
	}

	@Override
	public PermissionAttachment addAttachment(Plugin plugin, String name, boolean value) {
		return perm.addAttachment(plugin, name, value);
	}

	@Override
	public PermissionAttachment addAttachment(Plugin plugin) {
		return perm.addAttachment(plugin);
	}

	@Override
	public PermissionAttachment addAttachment(Plugin plugin, String name, boolean value, int ticks) {
		return perm.addAttachment(plugin, name, value, ticks);
	}

	@Override
	public PermissionAttachment addAttachment(Plugin plugin, int ticks) {
		return perm.addAttachment(plugin, ticks);
	}

	@Override
	public void removeAttachment(PermissionAttachment attachment) {
		perm.removeAttachment(attachment);
	}

	@Override
	public void recalculatePermissions() {
		perm.recalculatePermissions();
	}

	@Override
	public Set<PermissionAttachmentInfo> getEffectivePermissions() {
		return perm.getEffectivePermissions();
	}

	@Override
	public boolean isOp() { // TODO: Connect with TBMC acc
		return false;
	}

	@Override
	public void setOp(boolean value) { // TODO: Connect with TBMC acc
	}

	@Override
	public Server getServer() {
		return Bukkit.getServer();
	}

	@Override
	public String getName() {
		return name == null ? user == null ? "Discord user" : user.getDisplayName(DiscordPlugin.mainServer) : name;
	}

	@Override
	public Spigot spigot() {
		return new CommandSender.Spigot();
	}

}

File: src/main/java/buttondevteam/discordplugin/DiscordSender.scala
package buttondevteam.discordplugin

import discord4j.core.`object`.entity.User
import discord4j.core.`object`.entity.channel.MessageChannel
import org.bukkit.command.CommandSender
import org.bukkit.permissions.{PermissibleBase, Permission, PermissionAttachment, PermissionAttachmentInfo}
import org.bukkit.plugin.Plugin
import org.bukkit.{Bukkit, Server}
import reactor.core.scala.publisher.SMono

import java.util

class DiscordSender(user: User, channel: MessageChannel, pname: String) extends DiscordSenderBase(user, channel) with CommandSender {
    private val perm = new PermissibleBase(this)
    private val name: String = Option(pname)
        .orElse(Option(user).map(u => SMono(u.asMember(DiscordPlugin.mainServer.getId))
            .onErrorResume(_ => SMono.empty).blockOption()
            .map(u => u.getDisplayName)))
        .getOrElse("Discord user")

    def this(user: User, channel: MessageChannel) {
        this(user, channel, null)
    }

    override def isPermissionSet(name: String): Boolean = perm.isPermissionSet(name)

    override def isPermissionSet(perm: Permission): Boolean = this.perm.isPermissionSet(perm)

    override def hasPermission(name: String): Boolean = {
        if (name.contains("essentials") && !(name == "essentials.list")) return false
        perm.hasPermission(name)
    }

    override def hasPermission(perm: Permission): Boolean = this.perm.hasPermission(perm)

    override def addAttachment(plugin: Plugin, name: String, value: Boolean): PermissionAttachment = perm.addAttachment(plugin, name, value)

    override def addAttachment(plugin: Plugin): PermissionAttachment = perm.addAttachment(plugin)

    override def addAttachment(plugin: Plugin, name: String, value: Boolean, ticks: Int): PermissionAttachment = perm.addAttachment(plugin, name, value, ticks)

    override def addAttachment(plugin: Plugin, ticks: Int): PermissionAttachment = perm.addAttachment(plugin, ticks)

    override def removeAttachment(attachment: PermissionAttachment): Unit = perm.removeAttachment(attachment)

    override def recalculatePermissions(): Unit = perm.recalculatePermissions()

    override def getEffectivePermissions: util.Set[PermissionAttachmentInfo] = perm.getEffectivePermissions

    override def isOp = false

    override def setOp(value: Boolean): Unit = {
    }

    override def getServer: Server = Bukkit.getServer

    override def getName: String = name

    override def spigot = new CommandSender.Spigot
}